### DISCUSSION TITLE
How do you use getStaticProps in Layout component?

### DISCUSSION THREAD

**Discussion by varunsingh87** [vercel/next.js#16684] [Help] [answered]
I am making a newspaper website, and the newspaper columns are going to be stored in the database. These columns will appear on every page, so I want to put it in my `components/Layout.js`. However, `getStaticProps()` doesn't seem to get called in a component, only in a page. 

```
/**
 * @returns { Object } An object that contains a props object which contains an allPostsData property
 */
export async function getStaticProps() {
  const allColumnsData = getColumnsData();
  return {
    props: {
      columnsData: allColumnsData
    }
  }
}
```

Is it possible to get data at build time in a component, and if not, is there another way for me to do this?

**Comment by varunsingh87** [comment]
Update: I found [here](https://nextjs.org/docs/basic-features/data-fetching#only-allowed-in-a-page) that you can't have a component use getStaticProps() so how do I work around this?

**Comment by jamesmosier** [comment]
You would need to call your `getColumnsData` function in `getStaticProps` on every page and then you could pass that as a prop to your layout component.

Currently there is no way to globally hydrate data for an entire application, other than using `getInitialProps` in `_app.js` which unfortunately then disables automatic static optimization [per the docs](https://nextjs.org/docs/advanced-features/custom-app#caveats)

**Reply by IrfDev** [reply]
👍🏻

**Reply by insivika** [reply]
How do you then pass the prop up to the layout when the page is wrapped by the layout 🤔

**Reply by ljosberinn** [reply]
You can abstract this away:

```js
// _app.js

function App({ Component, pageProps }) {
  const withLayout = Component.withLayout ?? (page) => page;

  const { myCustomData, ,...rest } = pageProps;

  return withLayout(<Component {...rest} />, myCustomData); 
}

// page
function Page(props) {}

Page.withLayout = (page, myCustomData) => <Layout {...myCustomData}>{page}</Layout>

```

**Reply by timmysmalls** [reply]
Won't just using pageProps do the trick? Here's what I did in my `_app.js`:

```
export default ({ Component, pageProps }) => {
    return (
        <>  
        <Navbar {...pageProps.socialMediaSettings} />
        <Component {...pageProps} />
        <Footer />
        </> 
    );
};
```

So the socialMediaSettings are the global props I want to pass to my layout component. They just contain links to facebook and whatnot. I request them in every statically rendered page. So if I have my page query:

```
page(slug: $slug) {
    title
    subtitle
    image {
        src
    }
    whatever {
        else {
            ...
        }
    }
}
```

I'll wrap this in my wrapper `defaultPageQuery` function that always requests the necessary layout props alongside the original query:

```
export default query => {
    return`query ($slug: String) {
        socialMediaSettings {
            facebook
            twitter
            linkedin
        }
        ${query}
    }`;
```

Now every query that get wrapped with this function will also return all the info needed to render the layout. The `pageProps` in the `_app.js` will pass the props to the Layout component itself, as well as to the original component. I abstracted this further, but I think you get the point. There are probably more beautiful ways to do this, that do not involve all this string interpolation, but I'm old fashioned that way.

**Reply by ljosberinn** [reply]
If your Navbar is on every page, then sure! If you have multiple layouts, then not so much. Also you probably want to destructure `socialMediaSettings` from `pageProps` because its unlikely to be relevant within a page.

**Reply by mshshuvooo** [reply]
@jamesmosier How can I use getStaticProps in _app.js? Can you please give me an example? I tried many times but got undefined.

**Reply by ljosberinn** [reply]
@mshshuvooo the Next.js docs [explicitly mention](https://nextjs.org/docs/advanced-features/custom-app#caveats) this doesn't work.

See https://github.com/vercel/next.js/discussions/10949, https://github.com/vercel/next.js/discussions/11334, https://github.com/vercel/next.js/discussions/12439, https://github.com/vercel/next.js/discussions/12439 and https://github.com/vercel/next.js/issues/13225.

**Reply by molimat** [reply]
In my case, my work around was creating a getStaticProps in every page and populate a context (when it is empty) with the layout data. And then the layout can consume this context. I dont think it is beautiful nor flawless (like when the user just goes automatically to a dynamic link that I didnt fetch the layout data in getStaticProps due to performance matters, the layout wont be hydrated).

**Reply by muftisamiullah** [reply]
but getInitialProps has been deprecated? whats the work around? i need to get the menu items from an api how to i do that in a component rather than in a page.

**Reply by mshshuvooo** [reply]
> but getInitialProps has been deprecated? whats the work around? i need to get the menu items from an api how to i do that in a component rather than in a page.

I am having the same problem. Need a solution.

**Reply by handletastic** [reply]
This doesn't work particularly well if you want the Layout to use the revalidate feature from getStaticProps.
I have been trying ways to make this work but to no success.

In a nutshell my intent would be:
- to have a page that receives a layout
- for that layout to have a specified number of props that it requires
- for the layout props to receive these from a getStaticProps function so that they benefit from revalidate, ISR and that it'd be defined and fetched from a single place.

any ideas?

Thanks!

**Reply by abergenw** [reply]
@handletastic I suggest you take a look at [Next Page Layout](https://github.com/abergenw/next-page-layout). Using this simple lib it should be easy to achieve what you want.

**Reply by handletastic** [reply]
I have taken a look, thanks for the recommendation. Not going to lie, there is a bit of boilerplate and I would try to avoid having such heaviness in the codebase. Specifically the makeLayoutPage and makeLayout functions. Was hoping for a more elegant solution. But nonetheless, if this solves the issue I will gladly use it.
Also, how would testing packages integrate with the library to mock api responses and other types of tests?

Thanks, :)

**Comment by revskill10** [comment]
I think (still not tried yet) putting global data logic in `_document.js` is feasible answer till now. We still need to make a call in client-side to have update data.

**Reply by ptrhck** [reply]
I am trying to apply the concept of @ljosberinn but am struggeling a bit. Can someone help how to pass the props in this case:

__app.tsx_

```
import { CheckAuth } from "../auth/features"
import { Layout } from "../dismantling/features"
import { config } from "../dismantling/state"
import { NeoPage } from "../types"
import { CommonProviders } from "../utils"
import "../utils/i18n"
import "mapbox-gl/dist/mapbox-gl.css"
import type { AppProps /*, AppContext */ } from "next/app"

const NoCheck: React.FC = ({ children }) => <>{children}</>

function App({ Component, pageProps }: AppProps) {
  const neoPage = Component as NeoPage
  const LayoutComponent = neoPage.Layout || Layout

  // withAuthCheck is true by default. You have to exclude pages explicitly.
  const { withAuthCheck = true } = neoPage

  // Decide if we want to check auth or just render the page
  const CheckAuthComponent = withAuthCheck ? CheckAuth : NoCheck

  return (
    <CommonProviders pageProps={pageProps} overmindConfig={config}>
      <CheckAuthComponent>
        <LayoutComponent>
          <Component {...pageProps} />
        </LayoutComponent>
      </CheckAuthComponent>
    </CommonProviders>
  )
}

export default App
```

_Login.tsx_

```
import { LayoutUnauthorized } from "../design-system"
import { NeoPage } from "../types"


export const Login: NeoPage = () => {
  return (
    ...
  )
}

Login.Layout = LayoutUnauthorized
Login.withAuthCheck = false
```

So Layout has a property I would like getStaticProps to pass.

**Comment by abergenw** [comment]
I recently created a small lib which adds various functionality to layouts such as the possibility to define `getStaticProps` on your layouts: https://github.com/abergenw/next-page-layout

**Comment by wordanjood** [comment]
I decided to do a fetch to contentful api within the layout component. This meant I could get data I would from `getStaticProps` within the Layout component. Document to follow is [here](https://www.contentful.com/developers/docs/references/content-delivery-api/). The route to get all contentful entries is `https://cdn.contentful.com/spaces/${spaceId}/environments/${environmentId}/entries?access_token=${accessToken}`. 

I can't see anyway of returning just entries for specific content types so I simply just filtered the response by content id's. 

Hope this helps someone

**Comment by yassh** [comment]
# My solution

## JavaScript

**_app.js**

```jsx
const MyApp = ({ Component, pageProps }) => {
  const getLayout = Component.getLayout ?? ((page) => page)

  return getLayout(<Component {...pageProps} />, pageProps)
}

export default MyApp
```

**page**

```jsx
const Page = (props) => {
  /* ... */
}

Page.getLayout = (page, pageProps) => {
  <Layout columnsData={pageProps.columnsData}>
    {page}
  </Layout>
}

export default Page
```

## TypeScript

**_app.tsx**

```tsx
export type NextPageWithLayout<P = {}, IP = P> = NextPage<P, IP> & {
  getLayout?: (page: ReactElement, pageProps: P) => ReactNode
}

type AppPropsWithLayout = AppProps & {
  Component: NextPageWithLayout
}

const MyApp = ({ Component, pageProps }: AppPropsWithLayout) => {
  const getLayout = Component.getLayout ?? ((page) => page)

  return getLayout(<Component {...pageProps} />, pageProps)
}

export default MyApp
```

**page**

```tsx
const Page: NextPageWithLayout<Props> = (props) => {
  /* ... */
}

Page.getLayout = (page, pageProps) => {
  <Layout columnsData={pageProps.columnsData}>
    {page}
  </Layout>
}

export default Page
```

**Comment by dpodz** [comment]
Not sure if this is totally related, but I found a somewhat jank solution to using static props in a layout by just grabbing the props from the children. I'm sure I'm breaking some design patterns, but it works. But god I hate this solution, I really hope there's something better:

```tsx

const Page = ({
  // Props, etc
  assessmentIds,
}: {
  // Props, etc
  assessmentIds: CurriculumIds
}) => {
  // The whole page
}

export const PageLayout: React.FC<{}> = ({children}) => {
  const assessmentIds = (children as any).props.assessmentIds

  useEffect(() => {}, [(children as any).props])

  return (
    <UserRequired>
      {assessmentIds && (
        <AssessmentProvider
          chainId={assessmentIds.chainId as string}
          courseId={assessmentIds.courseId as string}
          moduleId={assessmentIds.moduleId as string}
        >
          {children}
        </AssessmentProvider>
      )}
    </UserRequired>
  )
}

Page.Layout = PageLayout
export default Page
```

**Comment by leerob** [comment, marked as answer]
**Update:** We’re excited to announce [Next.js 13](https://nextjs.org/13), which includes the new `app/` directory (beta) with support for nested layouts, colocation of data fetching, streaming, Server Components by default, and much more.

Inside the `app/` directory, there is a powerful new way to fetch data with React’s `use()` hook and the extended Web `fetch()` API. This allows you to colocate data fetching directly inside components for the most flexibility, including inside layouts.

You can leave feedback here: https://github.com/vercel/next.js/discussions/41745
